cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -1401,13 +1412,27 @@ private AppFabricServiceStatus stop(Id.Program identifier, ProgramType type) {
     }
   }
 
-  private void getRuns(HttpResponder responder, Id.Program programId, String status,
-                       long start, long end, int limit) {
+  private void getRuns(HttpResponder responder, Id.Program programId, String status, long start, long end, int limit) {
     try {
       try {
-        ProgramRunStatus runStatus = (status == null) ? ProgramRunStatus.ALL :
+        final ProgramType programType = programId.getType();
+        final ProgramRunStatus runStatus = (status == null) ? ProgramRunStatus.ALL :
           ProgramRunStatus.valueOf(status.toUpperCase());
-        responder.sendJson(HttpResponseStatus.OK, store.getRuns(programId, runStatus, start, end, limit));
+
+        // Need to validate whether the run records are reflecting the actual running state of the program.
+        List<RunRecord> runRecords = lifecycleService.validateRunRecords(
+          store.getRuns(programId, runStatus, start, end, limit), programId, programType);
```

Comment:
Yes, but we do not care for those that are is KILLED. The check is done for programs that at the execution time of this method is not actually running via call to the runtime service but the run records said it is running.

In the case of KILLED, the runtime service will tell that there is no program running for that id and hence the run records are not modified.
We do not handle the case for programs that are actually RUNNING but the run records state otherwise.

I shoule have just called the new method something like compareAndSetRunningRunRecord